File: emaalgotradingbot.ts
import * as fs from 'fs';

// Read data from CSV file
function readCSV(filename: string): number[] {
    const data: string[] = fs.readFileSync(filename, 'utf8').split('\n');
    const closePrices: number[] = [];

    for (const line of data) {
        const values: string[] = line.split(',');
        if (values.length >= 5) {
            closePrices.push(parseFloat(values[4])); // Assuming the close prices are in the 5th column (index 4)
        }
    }

    return closePrices;
}

// Exponential Moving Average calculation
function calculateEMA(prices: number[], period: number): number[] {
    const ema: number[] = [];
    const multiplier: number = 2.0 / (period + 1);
    let emaValue: number = prices[0];
    ema.push(emaValue);

    for (let i = 1; i < prices.length; i++) {
        emaValue = (prices[i] - emaValue) * multiplier + emaValue;
        ema.push(emaValue);
    }

    return ema;
}

// Remove NaN values from a list
function removeNaN(list: (number | null)[]): number[] {
    const result: number[] = [];
    for (const value of list) {
        if (value !== null) {
            result.push(value);
        }
    }
    return result;
}

// Calculate profit based on buy and sell prices
function calculateProfit(buyPrices: number[], sellPrices: number[]): number[] {
    const profit: number[] = [];
    const size: number = Math.min(buyPrices.length, sellPrices.length);
    for (let i = 0; i < size; i++) {
        profit.push(sellPrices[i] - buyPrices[i]);
    }
    return profit;
}

// Main function
function main() {
    // Read data from CSV file
    const closePrices: number[] = readCSV('BTC-USD.csv');

    // Exponential Moving Average calculation with period 30
    const ema30: number[] = calculateEMA(closePrices, 30);

    // Implement trading strategy
    const buySignals: (number | null)[] = [];
    const sellSignals: (number | null)[] = [];
    let flag: number = 0;
    let buyPrice: number = 0;

    for (let i = 0; i < closePrices.length; i++) {
        if (ema30[i] > closePrices[i] && flag === 0) {
            buySignals.push(closePrices[i]);
            sellSignals.push(null);
            buyPrice = closePrices[i];
            flag = 1;
        } else if (ema30[i] < closePrices[i] && flag === 1 && buyPrice < closePrices[i]) {
            sellSignals.push(closePrices[i]);
            buySignals.push(null);
            buyPrice = 0;
            flag = 0;
        } else {
            buySignals.push(null);
            sellSignals.push(null);
        }
    }

    // Print buy prices
    console.log('Bought at prices: ', removeNaN(buySignals));

    // Print sell prices

File: emaalgotradingbot.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class ExponentialMovingAverage {

    public static void main(String[] args) {
        // Read data from CSV file
        List<Double> closePrices = readCSV("BTC-USD.csv");

        // Exponential Moving Average calculation with period 30
        List<Double> ema30 = calculateEMA(closePrices, 30);

        // Implement trading strategy
        List<Double> buySignals = new ArrayList<>();
        List<Double> sellSignals = new ArrayList<>();
        int flag = 0;
        double buyPrice = 0;

        for (int i = 0; i < closePrices.size(); i++) {
            if (ema30.get(i) > closePrices.get(i) && flag == 0) {
                buySignals.add(closePrices.get(i));
                sellSignals.add(Double.NaN);
                buyPrice = closePrices.get(i);
                flag = 1;
            } else if (ema30.get(i) < closePrices.get(i) && flag == 1 && buyPrice < closePrices.get(i)) {
                sellSignals.add(closePrices.get(i));
                buySignals.add(Double.NaN);
                buyPrice = 0;
                flag = 0;
            } else {
                buySignals.add(Double.NaN);
                sellSignals.add(Double.NaN);
            }
        }

        // Print buy prices
        System.out.println("Bought at prices: " + removeNaN(buySignals));

        // Print sell prices
        System.out.println("Sold at prices: " + removeNaN(sellSignals));

        // Calculate and print profit
        List<Double> profit = calculateProfit(removeNaN(buySignals), removeNaN(sellSignals));
        System.out.println("Realized P/L: " + sumList(profit));

        // Visualization logic can be implemented here
    }

    // Read data from CSV file
    private static List<Double> readCSV(String filename) {
        List<Double> closePrices = new ArrayList<>();
        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = br.readLine()) != null) {
                String[] data = line.split(",");
                closePrices.add(Double.parseDouble(data[4])); // Assuming the close prices are in the 5th column (index 4)
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        return closePrices;
    }

    // Exponential Moving Average calculation
    private static List<Double> calculateEMA(List<Double> prices, int period) {
        List<Double> ema = new ArrayList<>();
        double multiplier = 2.0 / (period + 1);
        double emaValue = prices.get(0);
        ema.add(emaValue);
        for (int i = 1; i < prices.size(); i++) {
            emaValue = (prices.get(i) - emaValue) * multiplier + emaValue;
            ema.add(emaValue);
        }
        return ema;
    }

    // Remove NaN values from a list
    private static List<Double> removeNaN(List<Double> list) {
        List<Double> result = new ArrayList<>();
        for (Double value : list) {
            if (!Double.isNaN(value)) {
                result.add(value);
            }
        }
        return result;
    }

    // Calculate profit based on buy and sell prices
    private static List<Double> calculateProfit(List<Double> buyPrices, List<Double> sellPrices) {
        List<Double> profit = new ArrayList<>();
        int size = Math.min(buyPrices.size(), sellPrices.size());
        for (int i = 0; i < size; i++) {
            profit.add(sellPrices.get(i) - buyPrices.get(i));
        }
        return profit;
    }

    // Calculate sum of a list
    private static double sumList(List<Double> list) {
        double sum = 0;
        for (Double value : list) {
            sum += value;
        }
        return sum;
    }
}
